### PR TITLE
Rework association mapping hierarchy

### PR DESCRIPTION
**Pull request by greg0ire**
Please review  https://github.com/doctrine/orm/pull/10682 before this one.

- Each type is now either final, abstract or an interface. Inverse mappings are no longer represented by intermediary types, which means you no longer have this weird inheritance between owning and inverse. 
- The mappedBy attribute is no longer nullable and moved down the hierarchy, because it makes sense only for inverse side mappings.
- The inversedBy attribute is still nullable and also moved down the hierarchy, because it makes sense only for owning side mappings.
- Code common to ManyToOneAssociationMapping and OneToOneOwningSideMapping is de-duplicated and moved up the hierarchy
- Code inside ToManyInverseSideMapping and ToManyOwningSideMapping comes from a trait (:nauseated_face:) to avoid duplication.

# Before

## Classes

```mermaid
classDiagram
    AssociationMapping <|-- ToManyAssociationMapping
    ToManyAssociationMapping <|-- ManyToManyAssociationMapping
    ManyToManyAssociationMapping <|-- ManyToManyOwningSideMapping
    AssociationMapping <|-- ToOneAssociationMapping
    ToOneAssociationMapping <|-- ManyToOneAssociationMapping
    ToManyAssociationMapping <|-- OneToManyAssociationMapping
    ToOneAssociationMapping <|-- OneToOneAssociationMapping
    OneToOneAssociationMapping <|-- OneToOneOwningSideMapping
```

## Interfaces

```mermaid
classDiagram
    AssociationOwningSideMapping <|.. OneToOneOwningSideMapping
    AssociationOwningSideMapping <|.. ManyToManyOwningSideMapping
    AssociationOwningSideMapping <|.. ManyToOneAssociationMapping
```

# After

## Classes

```mermaid
classDiagram
    AssociationMapping <|-- InverseSideMapping
    AssociationMapping <|-- OwningSideMapping
    InverseSideMapping <|-- ToManyInverseSideMapping
    OwningSideMapping <|-- ToManyOwningSideMapping
    ToManyInverseSideMapping <|-- ManyToManyInverseSideMapping
    ToManyOwningSideMapping <|-- ManyToManyOwningSideMapping
    OwningSideMapping <|-- ToOneOwningSideMapping
    ToOneOwningSideMapping <|-- ManyToOneOwningSideMapping
    ToManyInverseSideMapping <|-- OneToManyAssociationMapping
    InverseSideMapping <|-- ToOneInverseSideMapping
    ToOneInverseSideMapping <|-- OneToOneInverseSideMapping
    ToOneOwningSideMapping <|-- OneToOneOwningSideMapping
```

## Interfaces

```mermaid
classDiagram
    ToManyAssociationMapping <|.. ManyToManyAssociationMapping
    ManyToManyAssociationMapping <|.. ManyToManyInverseSideMapping
    ManyToManyAssociationMapping <|.. ManyToManyOwningSideMapping
    ToOneAssociationMapping <|.. OneToOneAssociationMapping
    OneToOneAssociationMapping <|.. OneToOneInverseSideMapping
    OneToOneAssociationMapping <|.. OneToOneOwningSideMapping
```

## `ToManyAssociationMappingImplementation` Trait

```mermaid
classDiagram
    AssociationMapping <|-- InverseSideMapping
    AssociationMapping <|-- OwningSideMapping
    InverseSideMapping <|-- ToManyInverseSideMapping
    OwningSideMapping <|-- ToManyOwningSideMapping
    ToManyAssociationMappingImplementation <|-- ToManyInverseSideMapping
    ToManyAssociationMappingImplementation <|-- ToManyOwningSideMapping
```

### CODE REVIEW

**Comment by greg0ire** [comment]
@mpdude please review :)

File: lib/Doctrine/ORM/Mapping/ToManyAssociationMappingImplementation.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\ORM\Mapping;
+
+trait ToManyAssociationMappingImplementation
```
Comment (derrabus) [inline lib/Doctrine/ORM/Mapping/ToManyAssociationMappingImplementation.php:8]:
```php
/** @internal */
```

File: lib/Doctrine/ORM/Mapping/ToOneOwningSideMapping.php [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    /** @return list<string> */
+    public function __sleep(): array
+    {
+        $serialized = parent::__sleep();
+
+        $serialized[] = 'joinColumns';
+        $serialized[] = 'sourceToTargetKeyColumns';
+        $serialized[] = 'targetToSourceKeyColumns';
+
+        return $serialized;
```
Comment (derrabus) [inline lib/Doctrine/ORM/Mapping/ToOneOwningSideMapping.php]: Not a blocker, but you could make use of the spread operator here.

```suggestion
        return [
            ...parent::__sleep(),

            'joinColumns',
            'sourceToTargetKeyColumns',
            'targetToSourceKeyColumns',
        ];
```

File: lib/Doctrine/ORM/Mapping/OwningSideMapping.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    public string|null $inversedBy = null;
+
+    /** @return list<string> */
+    public function __sleep(): array
+    {
+        $serialized = parent::__sleep();
+
+        if ($this->inversedBy !== null) {
+            $serialized[] = 'inversedBy';
+        }
```
Comment (derrabus) [inline lib/Doctrine/ORM/Mapping/OwningSideMapping.php:24]: What's the rationale behind excluding the property if it's set to `null`? Micro-optimization of the serialized string?

Comment (greg0ire) [reply lib/Doctrine/ORM/Mapping/OwningSideMapping.php:24]: Yes, exactly. That what has been done for every other property. It was partially done in the past here: https://github.com/doctrine/orm/blob/9bc6f5b4ac6f1e7d4248b2efbd01a748782075bc/lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php#L970-L1071

Now, I do it at a deeper level.

**Review by derrabus — APPROVED** [review]

**Comment by mpdude** [comment]
Let's hope that future merge-ups will not be too much of a headache with these changes. 🤞🏻

**Comment by greg0ire** [comment]
Not much more than previous ones I think… I'm mostly reworking new classes here :slightly_smiling_face:

**Pull request merged**